### PR TITLE
Add Edit Submission Feature

### LINKED ISSUE #5: Edit submission data
Users should be able to edit the 'raw' submission data, e.g., to fix mistakes reported by enumerators, remove suspicious values highlighted during validation, etc. 

This should be possible through the platform - ideally through the view submissions page. The [qol-survey](https://github.com/stats4sd/qol-survey) platform has a basic version of this that we could use for now, although eventually, we will need a version with a better UX. 

To protect against mistakes during editing, and to enable a full audit of the research process, we should do 2 things: 

1. Store the 'raw' data as it comes directly from ODK, without any edits, and present a different version of the data for editing. 
2. Keep a history of the edits that are made to the data. We can use one of the following packages: 
   - https://laravel-auditing.com/guide/installation.html
   - https://github.com/VentureCraft/revisionable

### PR DESCRIPTION
**Pull request by dan-tang-ssd**
This PR is submitted to fix #5 

It is now WIP, this PR serves as status update.

### CODE REVIEW

**Comment by dan-tang-ssd** [comment]
I am using Drinks ODK form for testing in local env. 
I have commented some Tape ODK form specific columns so that the page can be showed without error.

By using KeyValue, JSON value can be showed in key value pair nicely.

---

Some issues need to be addressed:
1. Error occurred when Save record in Edit page, even nothing is updated. I checked that we have casted "content" attribute as array in Submission model... not sure why it generates error...
2. It is common to have many attributes in an ODK submission, it would be nicer to show Edit page in a separate page instead of popup modal.
3. As KeyValue field allows us to interact with one-dimensional JSON object, not sure if we can enhance it to detect and show array attribute as another KeyValue field. It seems that we need to sort it out because user may need to update a value inside a repeat group.

---

Screen shots:

image

image

image

image

**Comment by dave-mills** [comment]
> Error occurred when Save record in Edit page, even nothing is updated. I checked that we have casted "content" attribute as array in Submission model... not sure why it generates error...

My assumption is that this is the same reason as your point 3 - the fact that there is some nested JSON, and the KeyValue field expects all values to be strings, instead of arrays... (thus the "Must be of type ?string, array given" error)

> It is common to have many attributes in an ODK submission, it would be nicer to show Edit page in a separate page instead of popup modal.

Agreed - A separate Edit page would look better here. 

> As KeyValue field allows us to interact with one-dimensional JSON object, not sure if we can enhance it to detect and show array attribute as another KeyValue field. It seems that we need to sort it out because user may need to update a value inside a repeat group.

Also agreed! Definitely a vital part for editing these nested json submissions. If we can't do that, then we'll need a different approach (maybe presenting the EntityValue entries for editing?)

**Comment by dan-tang-ssd** [comment]
> > Error occurred when Save record in Edit page, even nothing is updated. I checked that we have casted "content" attribute as array in Submission model... not sure why it generates error...
> 
> My assumption is that this is the same reason as your point 3 - the fact that there is some nested JSON, and the KeyValue field expects all values to be strings, instead of arrays... (thus the "Must be of type ?string, array given" error)


Yes, your guessing is correct. This is due to the nested JSON.
I manually edit submission.content to an one-dimensional array, then it works well to update value in front end.

image

---

> > As KeyValue field allows us to interact with one-dimensional JSON object, not sure if we can enhance it to detect and show array attribute as another KeyValue field. It seems that we need to sort it out because user may need to update a value inside a repeat group.
> 
> Also agreed! Definitely a vital part for editing these nested json submissions. If we can't do that, then we'll need a different approach (maybe presenting the EntityValue entries for editing?)

I will have a more detail look for filament KeyValue. 
And also check for any third party component for this.

**Comment by dan-tang-ssd** [comment]
Good news. I found a filament plugins called JSON Column, which supports to edit multi-dimensional JSON value.

Filament JSON Column
https://filamentphp.com/plugins/valentin-morice-json-column

I performed testing with positive result.

The last part is to change Edit button to show in a separate page instead of a popup modal.

---

Screen shots:

image

image

**Comment by dan-tang-ssd** [comment]
I am failed change the Edit button to show Edit page in a separate page instead of a popup modal.
Error occurred for asking "team" relationship in Submission model.

I will work on removing previously saved entities, entity_values, custom tables records first, will get back to it later

Question: If popup modal works fine to edit a submission, why it does not work when we show the Edit page in a separate page...?

---

Screen shots:

image

image

**Comment by dave-mills** [comment]
@dan-tang-ssd You will need to change the tenant relationship name. Submissions do not have a relationship called 'team' -> which is what the error told you. So you must tell the Resource which relationship should be used for tenancy ('owner').

**Comment by dan-tang-ssd** [comment]
I think this PR is now ready for review.

Testing performed with positive result in local env.

---

This PR contains below changes:

1. app panel > ODK Form Management > click "Show Submissions" button of selected xlsform > click "Edit" button of a submission > Edit page will be showed as a popup modal.
2. In the Edit popup modal, odk_id is disabled to avoid accidental update. Submission content is showed in JSON Column editor directly for editing. JSON Column viewer is hidden to save additional click for user.
3. JSON Column editor comes with syntax check, any JSON error will be showed on screen immediately.
4. After saving the updated submission content, previsouly saved entities and entity_values records will be deleted.
5. It will trigger to re-handle the updated submission content, previously saved custom table records will be deleted. New entities, entity_values and custom table records will be created.

---

Screen shots:

image

image

**Comment by dan-tang-ssd** [comment]
One new finding...

The "Download Raw Survey Data" button will export submission data BEFORE user editing. (I guess it calls API to extract submission data from ODK central directly)

The "Download Raw Survey Data" button will export submission data AFTER user editing. (It extracts entities and entity_values record from local database)

We may need to separate them into two different buttons for clarity.

---

Screen shots:

image

image

File: app/Filament/App/Resources/XlsformResource/RelationManagers/SubmissionsRelationManager.php [hunk 1]
```diff
@@ -2,31 +2,38 @@
...
 
 class SubmissionsRelationManager extends RelationManager
 {
     protected static string $relationship = 'submissions';
 
     public function isReadOnly(): bool
     {
-        return false;$this->getSelectedTableRecords();
+        return false;
+        $this->getSelectedTableRecords();
```
Comment (dave-mills) [inline app/Filament/App/Resources/XlsformResource/RelationManagers/SubmissionsRelationManager.php]: Well, this wasn't supposed to be there in the dev branch! Deleting :D

**Review by dave-mills — APPROVED** [review]
This is a good MVP for editing the submission data through the platform.  The package + app logic for re-creating the entities makes sense to me. 

The UI is good enough to publish and demonstrate that we can edit things, but we will need to find a better way of presenting the submission data before field supervisors will be able to use it. I've added this as a new issue #53.

**Pull request merged**